Allow formatting question and form descriptions using markdown

susnux opened the pull request:
* Resolves: #489 and resolves #1253

This will allow users to use markdown (CommonMark) for formatting descriptions (both form and question).
See also this recording for example usage:

https://user-images.githubusercontent.com/1855448/198033394-49cbb74b-3ba0-4180-8660-5c17f4a76235.mp4

Chartman123 commented:
@susnux just a quick question: is there a reason why you moved the title handling to the mixin?

susnux commented:
> @susnux just a quick question: is there a reason why you moved the title handling to the mixin?

Because it is duplicated code / shared code between the create and submit view.
So I moved it to the mixin instead of copying for both views.

> I think that the css styles need some more love... at least currently for me the layout moves when I switch between edit and view mode...

What kind of move? If the input is just plain text there should be no difference, but of cause markdown and the rendered output might take different space and so there might be movement.

Chartman123 commented:
Yes that makes sense... But you also did a switch from textarea to input and textarea to NcRichContenteditable. So there's probably some other CSS applied to those elements

Chartman123 commented:
@susnux I think I found the source of the jumping: 

https://github.com/nextcloud/forms/blob/458bea1cc27b2c7e27e5baf86a77ac09df835e86/src/components/Questions/Question.vue#L373

I disabled this rule in the dev tools and the jumping between edit and view is gone... Could you please verify this (for one line descriptions) and put it into the PR?

gohrner commented:
This may also fix https://github.com/nextcloud/forms/issues/441 ?

Chartman123 commented:
> This may also fix #441 ?
> 
> Not sure if this PR also supports markdown in the individual answer options (checkmark / radio button descriptions)?

I think it makes more sense to keep the focus on the descriptions here.

susnux commented:
>Could you please verify this (for one line descriptions) and put it into the PR?

@Chartman123 I can verify this was the cause, I removed that rule and pushed it to this PR. No jumping here anymore :)

Chartman123 reviewed (approved):
So for me it's fine now :)

Perhaps @jotoeri will find some time to review as well

jotoeri commented on src/views/Create.vue:
```diff
@@ -38,30 +38,37 @@
 					type="text"
-					@input="onTitleChange" />
+					@input="onTitleChange">
+				<div v-else
+					id="form-title"
+					class="form-title"
+					v-text="formTitle" />
```
Please use the same element as in submit.
https://github.com/nextcloud/forms/blob/1a9445e00b78af3e3f59bd9ffe92ca8c56df1dd6/src/views/Submit.vue#L39-L41

jotoeri replied:
And please just insert the text with the brackets instead of the vue-directive, so we have common code across the app (and across most of nextcloud btw.)

jotoeri commented on src/views/Create.vue:
```diff
@@ -38,30 +38,37 @@
-			<textarea ref="description"
-				v-model="form.description"
+			<NcRichContenteditable v-if="edit || !form.title"
+				ref="description"
```
Should be unused now?

jotoeri commented on src/views/Create.vue:
```diff
@@ -38,30 +38,37 @@
 			<h2>
 				<label class="hidden-visually" for="form-title">{{ t('forms', 'Form title') }}</label>
-				<textarea id="form-title"
+				<input v-if="edit || !form.title"
```
This kills the multiline input of Form titles. The textarea was somewhen added on purpose.

susnux replied:
I think I do not understand the issue here, as your said the title must not be multi-line https://github.com/nextcloud/forms/pull/1394#discussion_r1060832519

susnux replied:
Oh... Did you mean that `textarea` allows the text to break, even without newline, if the text is too long?
Because the forms title seems to support two "lines" until it overflows.

My intention behind `<input>` was to unified the it with the title input of `Question.vue` which was always `<input>`.

But... I am sorry, simply misunderstood you comment :/

jotoeri replied:
> Oh... Did you mean that `textarea` allows the text to break, even without newline, if the text is too long?

Yes, thats it. 👍 The title should break automatically if it is too long, but people should not be able to create a title with `\n` newlines.

- Most probably it would be worth to adapt the Question-titles to also use textarea (Or the RichContent as you started with). The current state is more of a historical reason, that we earlier had inputs everywhere, but changed the title somewhen because of complaints. - But please, if you're motivated to change it, do it in a separate PR. 😉

Chartman123 replied:
@jotoeri Btw currently you can add newlines to the form title on main branch. However, they will not be shown in the submit view.

jotoeri replied:
Oh, really? 👀 Probably one more good reason to get to the Rich-Content Component in a next PR?!

jotoeri commented on src/components/Questions/Question.vue:
```diff
@@ -79,18 +79,16 @@
+					@update:value="onDescriptionChange" />
 
-				<!-- Do not wrap the following line between tags! `white-space:pre-line` respects `\n` but would produce additional empty first line -->
 				<!-- eslint-disable-next-line -->
```
Can you please replace that with the specific version? 
`<!-- eslint-disable-next-line vue/no-v-html -->`

jotoeri commented on src/components/Questions/Question.vue:
```diff
@@ -212,9 +218,9 @@ export default {
 
-		onDescriptionChange({ target }) {
+		onDescriptionChange(value) {
 			this.autoSizeDescription()
```
autosize has no effect now.

jotoeri commented on src/views/Create.vue:
```diff
@@ -38,30 +38,37 @@
-				@input="onDescChange" />
+				:maxlength="maxStringLengths.formDescription"
+				@input="saveDescription" />
+			<!-- eslint-disable-next-line -->
```
`<!-- eslint-disable-next-line vue/no-v-html -->`

jotoeri commented on src/views/Create.vue:
```diff
@@ -170,30 +184,19 @@ export default {
 
 	data() {
 		return {
-			maxStringLengths: loadState('forms', 'maxStringLengths'),
-
-			questionMenuOpened: false,
```
Why?

jotoeri commented on src/views/Submit.vue:
```diff
@@ -39,9 +39,8 @@
 				{{ formTitle }}
 			</h2>
-			<!-- Do not wrap the following line between tags! `white-space:pre-line` respects `\n` but would produce additional empty first line -->
 			<!-- eslint-disable-next-line -->
```
`<!-- eslint-disable-next-line vue/no-v-html -->`

jotoeri commented on src/views/Create.vue:
```diff
@@ -438,11 +432,13 @@ export default {
-			padding-bottom: 20px;
 			margin: 0px;
 			resize: none;
+			font-size: 100%;
+			min-height: unset;
+			line-height: 1.5em;
+			padding-bottom: 20px;
+			color: var(--color-text-maxcontrast) !important;
```
Christmas tree sorting??

jotoeri commented on src/mixins/ViewsMixin.js:
```diff
@@ -58,6 +59,25 @@ export default {
 	},
 
+	computed: {
+		/**
+		 * Return form title, or placeholder if not set
+		 *
+		 * @return {string}
+		 */
+		formTitle() {
+			if (this.form.title) {
+				return this.form.title
+			}
+			return t('forms', 'New form')
+		},
```
Can you remove it also from Results.vue then?

susnux replied:
done

jotoeri reviewed (changes requested):
- Newlines in the description get lost
- FormHead still jumps around between edit/view
- Having no border on the question description, the same should apply to the form description. Or the other way round, i fear we have to move to borders on all, since one otherwise does not see the mode change within the header.

susnux commented:
I will work through this points tomorrow.

> * Newlines in the description get lost

This is because simple newlines are soft breaks in markdown so this is expected behavior you will have to use double newlines or hardbreaks (backslash+newline or two space + newline).

But I think we will have to migrate old descriptions to markdown for this, I will also have a look at this topic tomorrow.

By the way I want to thank you for your work, your reviews are very in depth, sadly a rare thing on open source projects :)

susnux commented:
Sorry for the delay, had some busy weeks, but I have some time for this now :)

susnux commented on src/components/Questions/Question.vue:
```diff
@@ -189,6 +190,11 @@ export default {
 		},
 
+		computedDescription() {
+			const md = MarkdownIt()
```
better: per view

susnux commented on src/views/Create.vue:
```diff
@@ -422,9 +417,8 @@ export default {
 			padding: 0 16px;
-			border: none;
 		}
-		.form-title {
```
reverted

susnux commented:
@jotoeri I think I resolved all review notes :rocket:

Chartman123 commented on src/views/Create.vue:
```diff
@@ -403,41 +367,40 @@ export default {
-			resize: none;
+
+			&.readonly {
+				padding: 0px 15px 17px; // compensate border of editable content
```
Still some moving of the elements between view and edit mode for the form description. If you add the 1px padding to the top at least for me it stays in place 🙂 

```suggestion
				padding: 1px 15px 17px; // compensate border of editable content
```

susnux replied:
Thank you! Works :)

Chartman123 replied:
Did this disappear again? I can see a small movement when switching between the modes... This might also be related to the NcRichContenteditable as this component seems to have a thinner border than the other input fields. I created an issue for the lib: https://github.com/nextcloud/nextcloud-vue/issues/3635

susnux commented:
I think this ready to go now :)

jotoeri reviewed (commented):
@jancborchardt Could you also comment design-wise? With some distancing i'm not sure, what would be best.

Generally, @susnux, please try to keep the PR's simple. Don't mix changes, that are not necessarily related. Also just shifting around of Code without obvious reason is just annoying and hard to review, especially if there is much more and reasonable code on the PR, that needs to be distinguished. Better create separate small PR's, that can be quickly reviewed.

jotoeri reviewed (commented):
Ah -  and empty form-title gets lost on sidebar and submit currently.

![grafik](https://user-images.githubusercontent.com/47433654/210424302-946ac657-987f-4ec2-9e73-a721c131b4f2.png)

susnux commented on src/views/Results.vue:
```diff
@@ -198,15 +198,12 @@ export default {
 		 * @return {string}
 		 */
-		formTitle() {
-			if (this.form.title) {
-				return this.form.title
-			}
-			return t('forms', 'New form')
+		downloadUrl() {
+			return generateOcsUrl('apps/forms/api/v2/submissions/export/{hash}', { hash: this.form.hash })
```
Oh sorry this seems to be a artifact from rebasing

susnux commented on src/components/Questions/Question.vue:
```diff
@@ -100,16 +97,19 @@
 </template>
 
 <script>
-import { directive as ClickOutside } from 'v-click-outside'
```
Ok, good to known, fixed that :)

susnux commented on src/components/Questions/Question.vue:
```diff
@@ -100,16 +97,19 @@
+import { directive as ClickOutside } from 'v-click-outside'
+import { defineComponent } from 'vue'
+
+export default defineComponent({
```
Sure :)
BTW it is not needed if you do not use typescript, even with vue3, but it is quite useful for IDE like for IntelliSense.

Chartman123 commented on src/views/Create.vue:
```diff
@@ -403,41 +367,40 @@ export default {
-			margin-bottom: 0; // because the input field has enough padding
-		}
-
-		.form-title,
-		.form-desc,
```
Just saw this comment after hitting the send button on my other comment... For reference here's the issue from the vue lib again https://github.com/nextcloud/nextcloud-vue/issues/3635

susnux commented on src/views/Create.vue:
```diff
@@ -38,30 +38,29 @@
-					:minlength="0"
+				<NcRichContenteditable id="form-title"
+					:value.sync="form.title"
+					:multiline="true"
```
Reverted the switch to `NcRichContent` for the title, the also solved the issue with empty form names on the app navigation (`NcRichContent` always adds a newline to the value resulting in a not empty string but `'\n'`).

susnux commented:
> Ah - and empty form-title gets lost on sidebar and submit currently.

Fixed that, this was cause by `NcRichContent` which always added a newline to the value.

jotoeri reviewed (changes requested):
Did you test this? Did you check the diff here?
- Title looks completely different on create.vue
- Distance between Desc. & Title different between Create and Submit

susnux commented:
> Did you test this? Did you check the diff here?
> 
>     * Title looks completely different on create.vue
>     * Distance between Desc. & Title different between Create and Submit

Of cause I tested this, I even took screenshots to pixel compare them (checking if there is any jumping between edit edit-view and submit).
But I did not spot any difference:

| Submit | Create read | Create edit |
|-------|---------|--------|
| ![submit view](https://user-images.githubusercontent.com/1855448/211639688-40caebe1-9df4-44b0-9cd5-948e608b56ac.png) | ![create view readonly](https://user-images.githubusercontent.com/1855448/211639686-c0afdf22-607d-488d-bb0d-b66578ae926a.png) | ![create view edit](https://user-images.githubusercontent.com/1855448/211639679-b1bd18ee-0271-408a-9df7-428a356a4533.png) |

jotoeri commented:
🤔 I now checked a second time and i think this is how it looks, if i check out the current PR... Can you double-check, if you pushed all commits or so?

| Submit | Create Read | Create Edit
|---|---|---|
|![grafik](https://user-images.githubusercontent.com/47433654/211642450-71b9ada9-03e3-4481-9aac-b6b79dcccc71.png)|![grafik](https://user-images.githubusercontent.com/47433654/211642589-a250914e-b9a2-40c8-adfa-dc6de34374f1.png)|![grafik](https://user-images.githubusercontent.com/47433654/211642853-0f1491b0-c429-410c-b11d-b713ff67e559.png)|

susnux commented:
> thinking I now checked a second time and i think this is how it looks, if i check out the current PR... Can you double-check, if you pushed all commits or so?

Everything up-to-date, but that looks like it uses a different font size within the input, which is not changed by the CSS, so this might be browser related (I use Firefox 108) or maybe caching?
Nevertheless as I revert the change to `input` this should be "fixed" then.

susnux commented:
I rebased this branch and reverted the textarea change to fix the issues mentioned above.
I checked also there is no jumping or other visual issues anymore.

| Submit | Create | Create edit | Create edit questions |
|------------|-----------|------------------|---------------------------------|
| ![screenshot submit view](https://user-images.githubusercontent.com/1855448/213437749-3f3f1db2-8521-4c9b-beee-a47530a6991d.png) | ![screenshot create view](https://user-images.githubusercontent.com/1855448/213437755-3d1e16a3-5a2c-4a3b-840a-05b7eb0be021.png) | ![screenshot create view editing header](https://user-images.githubusercontent.com/1855448/213437761-da710194-826c-4e68-8e72-53a764c4fe60.png) | ![screenshot create view editing questions](https://user-images.githubusercontent.com/1855448/213437767-4189b667-dd41-4dc9-8498-568571f61164.png) |

Chartman123 commented:
@susnux: Just a little thing that I saw right now: In your question edit screenshot the textboxes for the descriptions/options are a little bit wider than the title. I don't know if this was also before, but perhaps you can have a look at this :)

susnux commented:
> @susnux: Just a little thing that I saw right now: In your question edit screenshot the textboxes for the descriptions/options are a little bit wider than the title. I don't know if this was also before, but perhaps you can have a look at this

Sure! Fixed the border alignment, it now looks like this:
![screenshot question borders aligned](https://user-images.githubusercontent.com/1855448/213448114-86d1ddfe-ffd7-49be-a358-45044b9902a5.png)

Chartman123 commented:
@susnux @jotoeri Did some review of the application in the browser. Couldn't find any unwanted differences to the current main branch. So from my side everything should be ok. If there is nothing more codewise, we can give it a go :)

Chartman123 commented on package-lock.json:
```diff
@@ -2626,15 +2627,15 @@
 		"node_modules/@typescript-eslint/eslint-plugin": {
-			"version": "5.48.1",
-			"resolved": "https://registry.npmjs.org/@typescript-eslint/eslint-plugin/-/eslint-plugin-5.48.1.tgz",
-			"integrity": "sha512-9nY5K1Rp2ppmpb9s9S2aBiF3xo5uExCehMDmYmmFqqyxgenbHJ3qbarcLt4ITgaD6r/2ypdlcFRdcuVPnks+fQ==",
+			"version": "5.48.2",
+			"resolved": "https://registry.npmjs.org/@typescript-eslint/eslint-plugin/-/eslint-plugin-5.48.2.tgz",
+			"integrity": "sha512-sR0Gja9Ky1teIq4qJOl0nC+Tk64/uYdX+mi+5iB//MH8gwyx8e3SOyhEzeLZEFEEfCaLf8KJq+Bd/6je1t+CAg==",
```
Are these changes needed here? Also valid for the other related changes in here...

jotoeri replied:
Ah, and i personally dont mind too much on this, as its only on the lockfile here... Its indeed not related, but Dependabot would anyways change it with the next run...

Chartman123 reviewed (approved):
One more comment on the package-lock.json. Rest seems fine

jotoeri commented:
- Now its on the left side and even worse?! 
![grafik](https://user-images.githubusercontent.com/47433654/213556697-453f76d8-5257-4bb5-89cc-6893a0e30524.png)
- From empty Title, when entering the first char, the title input changes size for me. @Chartman123 can you confirm this?
- I think Form title and description should **always** change edit-mode together. Currently, if the form title is empty, the description disappears. That looks strange imo.
- If there's no form-description, the Title and the info-message are now quite far apart. In fact 20px more, than without this PR. Can we get that down again? I think this doesn't necessarily be pixel-perfect between create.vue and submit.vue, but more important it should look good on submit.vue.

@Chartman123 @susnux  any opinions?

Chartman123 commented:
> * Now its on the left side and even worse?!

That's basically the same in main and there since Forms 3 iirc



> * From empty Title, when entering the first char, the title input changes size for me. @Chartman123 can you confirm this?

No, can't reproduce here.

> * I think Form title and description should **always** change edit-mode together. Currently, if the form title is empty, the description disappears. That looks strange imo.

Yes, same here... didn't try this before, but yes: should be linked

> * If there's no form-description, the Title and the info-message are now quite far apart. In fact 20px more, than without this PR. Can we get that down again? I think this doesn't necessarily be pixel-perfect between create.vue and submit.vue, but more important it should look good on submit.vue.

Confirm, I think the description element should be condtionally shown only if there is some text

susnux commented:
I fall into line :)

> * Now its on the left side and even worse?!

I can remove this but I think this would be better placed in a follow up PR as it is unrelated, but I do not think this is bad as it makes clear what is the question header and what the body.

> * I think Form title and description should **always** change edit-mode together. Currently, if the form title is empty, the description disappears. That looks strange imo.

I agree, fixed that.

> * If there's no form-description, the Title and the info-message are now quite far apart. In fact 20px more, than without this PR. Can we get that down again? I think this doesn't necessarily be pixel-perfect between create.vue and submit.vue, but more important it should look good on submit.vue.

Same, pushed a commit which fixes this :)

jotoeri commented:
> Now its on the left side and even worse?!
Well, 'unrelated' depends - you increased the offset from 2px to 4px in this PR, instead of aligning the description to the existing field...
I'm fine with having a clear difference between question and options, but currently it looks more like a bug imo. So i'd say, @susnux  please clarify with @jancborchardt how it should like and eventually fix it in a separate PR.

> From empty Title, when entering the first char, the title input changes size for me.
That's strange... I still see it on my Firefox 109 but not on edge. 🤔

jotoeri reviewed (approved):
🤷‍♂️

jotoeri commented:
@susnux wanna rebase & squash the commits here? Then we can merge it.

susnux commented:
> wanna rebase & squash the commits here? Then we can merge it.

Nice you released the hotfix update :)

hanserasmus commented:
Are these changes included in version 3.0.4? I have 3.0.4 installed, and none of the above Markdown stuff works for me. Trying to add a URL to a question and/or description and it displays the raw text, not MD formatted.

susnux commented:
No this will be included with upcoming 3.1.0 release.

hanserasmus commented:
Thank you @susnux. Is there a place where I can watch the project timeline? I need this urgently for a project, so if needed be, I will build the app from source code to have this capability. But don't want to build from source if 3.1.0 is imminent.

susnux commented:
I do not think there is a fixed schedule, but I am not into that topic, maybe @jotoeri can help.

jotoeri commented:
Yes, there generally is no fixed release schedule for forms. But we anyways have to release for NC26 within the next 3 days. Maybe i'll find some time later, otherwise tomorrow.

hanserasmus commented:
Thank you @jotoeri . I saw the release, updated, and it broke. :-( Bug report filed: https://github.com/nextcloud/forms/issues/1528
Thank you still for the quick release! This will definitely help me once I get it working again.
